src/main/java/net/sf/jabref/logic/util/io/FileFinder.java

```diff
@@ -1,43 +1,42 @@
 package net.sf.jabref.logic.util.io;
 
 import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
 import java.util.Collection;
 import java.util.HashSet;
+import java.util.Objects;
 import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
 
 public class FileFinder {
 
-    public static Set<File> findFiles(Collection<String> extensions, Collection<File> directories) {
-        Set<File> result = new HashSet<>();
+    private static final Log LOGGER = LogFactory.getLog(FileFinder.class);
 
-        for (File directory : directories) {
-            result.addAll(FileFinder.findFiles(extensions, directory));
-        }
 
-        return result;
-    }
+    public static Set<File> findFiles(Collection<String> extensions, Collection<File> directories) {
+
+        Objects.requireNonNull(directories, "Directories must not be null!");
+        Objects.requireNonNull(extensions, "Extensions must not be null!");
 
-    private static Collection<? extends File> findFiles(Collection<String> extensions, File directory) {
         Set<File> result = new HashSet<>();
+        for (File directory : directories) {
 
-        File[] children = directory.listFiles();
-        if (children == null) {
-            return result; // No permission?
-        }
+            try (Stream<File> files = Files
+                    .find(directory.toPath(), Integer.MAX_VALUE,
+                            (path, attr) -> !Files.isDirectory(path)
```

Comment:
I can maybe split the Bi-Predicate lambda in a new function..
The reason for streams is that is lazily walking the file system and increases performance for huge directories.